File: solutions/c/yacht/1/yacht.c
#include "yacht.h"

#define DIE_FACES 6
#define NUM_DICE 5

int score(dice_t dice, category_t category) {
    int faces[DIE_FACES] = {0};
    int score = 0;
    int non_zero = 0;

    for (int i = 0; i < NUM_DICE; i++) {
        if (faces[dice.faces[i] - 1] == 0) non_zero++;
        faces[dice.faces[i] - 1]++;
        score += dice.faces[i];
    }

    if (ONES <= category && category <= SIXES)
        return faces[category] * (category + 1);
    else if (category == YACHT && non_zero == 1)
        return 50;
    else if (category == FULL_HOUSE) {
        for (int i = 0; i < NUM_DICE; i++)
            if (non_zero != 2 || (faces[dice.faces[i] - 1] != 2 &&
                                  faces[dice.faces[i] - 1] != 3))
                return 0;
        return score;
    } else if (category == FOUR_OF_A_KIND) {
        for (int i = 0; i < NUM_DICE; i++)
            if (faces[dice.faces[i] - 1] == 4 || faces[dice.faces[i] - 1] == 5)
                return 4 * (dice.faces[i]);
    } else if (non_zero == 5 &&
               ((category == LITTLE_STRAIGHT && faces[DIE_FACES - 1] == 0) ||
                (category == BIG_STRAIGHT && faces[0] == 0)))
        return 30;
    else if (category == CHOICE)
        return score;

    return 0;
}

File: solutions/python/yacht/1/yacht.py
# Score categories.
# Change the values as you see fit.
from collections import Counter


YACHT = 0
ONES = 1
TWOS = 2
THREES = 3
FOURS = 4
FIVES = 5
SIXES = 6
FULL_HOUSE = 7
FOUR_OF_A_KIND = 8
LITTLE_STRAIGHT = 9
BIG_STRAIGHT = 10
CHOICE = 11


def score(dice, category):
    if category == YACHT:
        return 50 if all(d == dice[0] for d in dice) else 0
    if ONES <= category <= SIXES:
        return category * dice.count(category)
    if category == FULL_HOUSE:
        return sum(dice) if all(count == 2 or count == 3 for _, count in Counter(dice).items()) else 0
    if category == FOUR_OF_A_KIND:
        counts = Counter(dice)
        return sum(d * 4 for d, count in counts.items() if count >= 4) if all(count == 1 or count >= 4 for _, count in counts.items()) else 0
    if category == LITTLE_STRAIGHT:
        return 30 if sorted(dice) == [1,2,3,4,5] else 0
    if category == BIG_STRAIGHT:
        return 30 if sorted(dice) == [2,3,4,5,6] else 0

    return sum(dice)
